### PR TITLE
bluetooth: tester: MCP Client tests

### PR DESCRIPTION
**Pull request by piotrnarajowski**
Add support for Media Control Client tests

### CODE REVIEW

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 1]
```diff
@@ -0,0 +1,1292 @@
...
+		search_items.len += sizeof(scp_cmd.len);
+
+		memcpy(&search_items.search[search_items.len], &scp_cmd.type,
+		       sizeof(scp_cmd.type));
+		search_items.len += sizeof(scp_cmd.type);
+
+		memcpy(&search_items.search[search_items.len], &scp_cmd.param,
+		       sizeof(scp_cmd.param));
+		search_items.len += strlen(scp_cmd.param);
+	}
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_mcp.c:1133]: Missing validation that `search_items.len <= ARRAY_SIZE(search_items.search)`.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 2]
```diff
@@ -0,0 +1,1292 @@
...
+	uint8_t param[SEARCH_LEN_MAX];
+
+	net_buf_simple_init(rx_ev_buf, 0);
+
+	ev = net_buf_simple_add(rx_ev_buf, sizeof(*ev));
+
+	bt_addr_le_copy(&ev->address, bt_conn_get_dst(conn));
+
+	ev->att_status = err;
+	ev->param_len = (uint8_t)search->search[0];
```
[the remark this answers is not in the document]
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp_mcp.c:260]: fixed

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 3]
```diff
@@ -0,0 +1,1292 @@
...
+	ev->search_type = search->search[1];
+	strcpy(param, search->search + 2);
```
[the remark this answers is not in the document]
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp_mcp.c]: fixed

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 4]
```diff
@@ -0,0 +1,1307 @@
...
+	LOG_DBG("MCC Segments Object ID read cb (%d)", err);
+
+	btp_send_current_track_obj_id_ev(conn, err ? 1 : 0, id);
+}
+
+static void mcc_current_track_obj_id_set_cb(struct bt_conn *conn, int err, uint64_t id)
+{
+	LOG_DBG("MCC Segments Object ID set cb (%d)", err);
+
+	btp_send_current_track_obj_id_ev(conn, err ? 1 : 0, id);
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_mcp.c:454]: `btp_send_current_track_obj_id_ev` and other claim the the `err` value is an ATT error code, but these are clearly always either `0` or `1` and not an ATT status.

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 5]
```diff
@@ -0,0 +1,1307 @@
...
+		       sizeof(scp_cmd.type));
+		search_items.len += sizeof(scp_cmd.type);
+
+		memcpy(&search_items.search[search_items.len], &scp_cmd.param,
+		       sizeof(scp_cmd.param));
+		search_items.len += strlen(scp_cmd.param);
+
+		if (search_items.len > ARRAY_SIZE(search_items.search)) {
+			return BTP_STATUS_FAILED;
+		}
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_mcp.c:1147]: You need to check this before you access `search_items.search[search_items.len]`, and also in the above `if clause`

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 6]
```diff
@@ -0,0 +1,1307 @@
...
+
+	ev->att_status = att_status;
+	ev->param_len = (uint8_t)search->search[0];
+
+	if (ev->param_len > (SEARCH_LEN_MAX - 1)) {
+		return;
+	}
+
+	ev->search_type = search->search[1];
+	strcpy(param, &search->search[2]);
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_mcp.c:267]: Are we sure that `param` is always NULL terminated?

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 7]
```diff
@@ -0,0 +1,1307 @@
...
+	net_buf_simple_init(rx_ev_buf, 0);
+
+	ev = net_buf_simple_add(rx_ev_buf, sizeof(*ev));
+
+	bt_addr_le_copy(&ev->address, bt_conn_get_dst(conn));
+
+	ev->att_status = att_status;
+	ev->param_len = (uint8_t)search->search[0];
+
+	if (ev->param_len > (SEARCH_LEN_MAX - 1)) {
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp_mcp.c:262]:
```suggestion
	if (ev->param_len > (SEARCH_LEN_MAX - sizeof(ev->param_len)) {
```

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/tester/src/btp_mcp.c (hunk 4 above)
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp_mcp.c:454]: Right, I'll change them to 'status'

Comment (Thalley) [reply tests/bluetooth/tester/src/btp_mcp.c:454]: What does the values 0 and 1 represent? BTP_STATUS_FAILED?

Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp_mcp.c:454]: Yes

Comment (Thalley) [reply tests/bluetooth/tester/src/btp_mcp.c:454]: The please use those :)

File: tests/bluetooth/tester/src/btp_mcp.c (hunk 6 above)
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp_mcp.c:268]: Fixed. Now we make sure it's NULL terminated before sending command to Search Control Point.

**Review by Thalley — CHANGES_REQUESTED** [review]
Minor comments, otherwise LGTM

File: tests/bluetooth/tester/src/btp/btp_mcp.h [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+#define BTP_MCP_CURRENT_GROUP_OBJ_ID_READ	0x0d
+struct btp_mcp_current_group_obj_id_read_cmd {
+	bt_addr_le_t address;
+} __packed;
+
+#define BTP_MCP_CURRENT_GROUP_OBJ_ID_SET	0x0e
+struct btp_mcp_current_group_obj_id_set_cmd {
+	bt_addr_le_t address;
+	uint64_t id;
```
Comment (MariuszSkamra) [inline tests/bluetooth/tester/src/btp/btp_mcp.h:81]: uint8_t id[3];

File: tests/bluetooth/tester/src/btp_mcp.c [hunk 9]
```diff
@@ -0,0 +1,1320 @@
...
+	}
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static uint8_t mcp_set_current_group_obj_id(const void *cmd, uint16_t cmd_len, void *rsp,
+					    uint16_t *rsp_len)
+{
+	const struct btp_mcp_current_group_obj_id_set_cmd *cp = cmd;
+	uint64_t id = sys_le64_to_cpu(cp->id);
```
[the remark this answers is not in the document]
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp_mcp.c]: It wasn't intentional, I missed that. Now all OTS OBJ IDs are 48-bit.

File: tests/bluetooth/tester/src/btp/btp_mcp.h [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+	bt_addr_le_t address;
+	uint8_t status;
+	uint8_t id[6];
+} __packed;
+
+#define BTP_MCP_NEXT_TRACK_OBJ_ID_EV		0x86
+struct btp_mcp_next_track_obj_id_ev {
+	bt_addr_le_t address;
+	uint8_t status;
+	uint8_t id[6];
```
Comment (Thalley) [inline tests/bluetooth/tester/src/btp/btp_mcp.h:193]: Consider using 
```suggestion
	uint8_t id[BT_OTS_OBJ_ID_SIZE];
```

For these

**Review by Thalley — DISMISSED** [review]

File: tests/bluetooth/tester/src/btp/btp_mcp.h (hunk 10 above)
Comment (piotrnarajowski) [reply tests/bluetooth/tester/src/btp/btp_mcp.h]: Changed

**Review by Thalley — DISMISSED** [review]
Minor comments but OK as is

**Review by Thalley — APPROVED** [review]

**Review by sjanc — APPROVED** [review]

**Review by MariuszSkamra — APPROVED** [review]

**Pull request merged**